app/src/main/kotlin/com/wire/android/util/permission/RecordAudioRequestFlow.kt

```diff
@@ -0,0 +1,54 @@
+package com.wire.android.util.permission
+
+import android.content.Context
+import android.net.Uri
+import androidx.activity.compose.ManagedActivityResultLauncher
+import androidx.activity.compose.rememberLauncherForActivityResult
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.compose.ui.platform.LocalContext
+import com.wire.android.util.extension.checkPermission
+
+@Composable
+fun rememberRecordAudioRequestFlow(
+    shouldPersistUri: (Boolean) -> Unit,
+    onPermissionDenied: () -> Unit,
+    onAudioRecordedUri: Uri
+): RecordAudioRequestFlow {
+    val context = LocalContext.current
+
+    val requestPermissionLauncher: ManagedActivityResultLauncher<Array<String>, Map<String, @JvmSuppressWildcards Boolean>> =
+        rememberLauncherForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { permissions ->
+            val allPermissionGranted = permissions.values.all { true }
+            if (allPermissionGranted) {
+                // TODO: launch record audio flow
+            } else {
+                onPermissionDenied()
+            }
+        }
+
+    return remember {
+        RecordAudioRequestFlow(context, requestPermissionLauncher)
+    }
+}
+
+class RecordAudioRequestFlow(
+    private val context: Context,
+    private val audioRecordPermissionLauncher: ManagedActivityResultLauncher<Array<String>, Map<String, @JvmSuppressWildcards Boolean>>
+) {
+    fun launch() {
+        if (context.checkPermission(android.Manifest.permission.RECORD_AUDIO) &&
+            context.checkPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
+        ) {
+            // TODO: launch record audio flow
+        } else {
+            audioRecordPermissionLauncher.launch(
+                arrayOf(
+                    android.Manifest.permission.RECORD_AUDIO,
+                    android.Manifest.permission.WRITE_EXTERNAL_STORAGE
```

Comment:
Would we really need to save the audio recordings on external storage? or should we proceed as with the image sharing, storing them temporarily in internal storage no matter whether they are picked from gallery or taken by the camera?

Reply:
I think probably we'll take the same approach when implementing the specific sharing features.  We'll have to create a temp file for every kind of attachment and just overwrite that when we want to send the N+1 audio file, for example.

Changed to `READ_EXTERNAL_STORAGE` ⚡